e2e/benchmarks/services_test.go

```diff
@@ -0,0 +1,247 @@
+	defer srv.Stop()
+	path, cleanup := setupServiceInstances(b, srv, N)
+	defer cleanup() // comment out if you want to keep the config files
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		require.NoError(b, runSyncOnce(path))
+	}
+}
+
+// Benchmark N Services
+func Benchmark1Service(b *testing.B) {
+	benchmarkServices(b, 1)
+}
+func Benchmark100Service(b *testing.B) {
+	benchmarkServices(b, 100)
+}
+func Benchmark1000Service(b *testing.B) {
+	benchmarkServices(b, 1000)
+}
+func benchmarkServices(b *testing.B, N int) {
+	srv := testutils.NewTestConsulServerHTTPS(b, "../../testutils")
+	defer srv.Stop()
+	path, cleanup := setupMultiServices(b, srv, N)
+	defer cleanup() // comment out if you want to keep the config files
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		require.NoError(b, runSyncOnce(path))
+	}
+}
+
+// ----------------------------------------------------------------------------
+
+// Run CTS in once mode from the cli level
+func runSyncOnce(configPath string) error {
+	cli := command.NewCLI(os.Stdout, os.Stderr)
+	args := []string{"cts", "-once", fmt.Sprintf("--config-file=%s", configPath)}
+	exitCode := cli.Run(args)
+	errstr := "Exit Code Error: %s (%d)\n"
+	switch exitCode {
+	case command.ExitCodeOK:
+		return nil
+	case command.ExitCodeError:
+		return fmt.Errorf(errstr, "ExitCodeError", exitCode)
+	case command.ExitCodeInterrupt:
+		return fmt.Errorf(errstr, "ExitCodeInterrupt", exitCode)
+	case command.ExitCodeRequiredFlagsError:
+		return fmt.Errorf(errstr, "ExitCodeRequiredFlagsError", exitCode)
+	case command.ExitCodeParseFlagsError:
+		return fmt.Errorf(errstr, "ExitCodeParseFlagsError", exitCode)
+	case command.ExitCodeConfigError:
+		return fmt.Errorf(errstr, "ExitCodeConfigError", exitCode)
+	case command.ExitCodeDriverError:
+		return fmt.Errorf(errstr, "ExitCodeDriverError", exitCode)
+	default:
+		return fmt.Errorf(errstr, "UNKNOWN", exitCode)
+	}
+}
+
+// Sets up consul and config to test setup with N service instances
+func setupServiceInstances(t testing.TB, srv *testutil.TestServer, N int) (string, func() error) {
+	services := make([]string, 1)
+	for _, s := range generateServices(N, true) {
+		testutils.RegisterConsulService(t, srv, s, testutil.HealthPassing)
+		services[0] = s.Name
+	}
+	return makeConfig(t, services, srv.HTTPAddr, false)
+}
+
+// add generated services as separte services
+func setupMultiServices(t testing.TB, srv *testutil.TestServer, N int) (string, func() error) {
+	services := make([]string, N)
+	for i, s := range generateServices(N, false) {
+		testutils.RegisterConsulService(t, srv, s, testutil.HealthPassing)
+		services[i] = s.Name
+	}
+	return makeConfig(t, services, srv.HTTPSAddr, true)
+}
+
+// generate service instance entries
+func generateServices(n int, instances bool) []testutil.TestService {
+	baseport := 30000
+	services := make([]testutil.TestService, n)
+	for i := 0; i < n; i++ {
+		name, id := fmt.Sprintf("svc_%d", i), "instance"
```

Comment:
Running the `BenchmarkNService` benchmarks locally, the `terraform.tfvars.tmpl` contains querying the N services, however I'm noticing that the `terraform.tfvars` file is not rendered with the expected services. For the 100 service benchmark, only `svc_099` renders in the tfvars file. For 1000 services I'm seeing an empty `services = { }`

The service ID is unique across all services for an agent. I suspect that `id = "instance"` is overriding the service registration instead of registering unique N services. Could you look into why the `tfvars` is not rendering as expected with 1 instance for N services?

Reply:
Interesting... I'll need to look into this one. :thinking:

Reply:
I think there's a race for the `services = {}` vs. the single entry. I initially only saw `services = {}` even with `100S` unless I put a sleep in between the population code and the CTS run.

So I need to figure that out as well.

Reply:
Interesting, a potential race and maybe the instance iD isn't an issue then 🤔

Reply:
No... not the race I thought it was as the sleep doesn't fix it reliably. Even with a long sleep it sometimes works and sometimes doesn't. I'm going to dig more into it today.

Reply:
Since the  100Instances test works as it should, I tried changing the 100Services test (the one having the problem) so each Service used had unique names and ids (previously only name was unique) as this was one of the differences from the 100Instances. The result it that I get a terraform.tfvars that usually contains 1-3 entries, with all their names/ids being early entries (they are numbered 1-100) so far always being <30.

I also added a call to the consul server to print out all services after the initialization but before the CTS run is triggered. All 100 services are registered.

I'm working to check the health status of those 100 services now.

Reply:
Printed out all the health checks for the services and they are all present via that endpoint as well and all passing. Again, this was printed after the seeding of test data into consul but before the CTS run. So that data is all present and good in Consul prior to the CTS run.

Might be an actual bug.